### PR TITLE
bridge: Don't require mmap for fsread1 channel

### LINKED ISSUE #5597: base: Unable to read non-mappable files
Some files can't be mapped, but we would still like to be able to read them via `cockpit.file`.

Example: Trying to read `/sys/kernel/kexec_crash_size` will result in the error
```
couldn't map: Failed to map fd: mmap() failed: No such device
```
https://github.com/cockpit-project/cockpit/blob/5196b70887fd89e0b4442ff55827fb7a71dfe14a/pkg/users/local.js#L86

We should probably provide a fallback solution here.

### PR DESCRIPTION
**Pull request by mvollmer**
This just slurps in the whole file, but all big files hopefully support mmap.

Fixes #5597

### CODE REVIEW

**Comment by mvollmer** [comment]
@dperpeet, check this out.  Do you also need cockpit.file.watch() to work?  It just might, but it's harder to write a unit test for that I guess.

**Comment by mvollmer** [comment]
> It just might, but it's harder to write a unit test for that I guess.

Hmm , maybe we can wrap mmap and let it fail exactly as we want.

**Comment by mvollmer** [comment]
>> It just might, but it's harder to write a unit test for that I guess.

> Hmm , maybe we can wrap mmap and let it fail exactly as we want.

Ouch, I was confused.  Watching files in /sys has nothing to do with mmap, but afaik they might require some special use of select.  Or not.  Just ignore me.

File: src/bridge/cockpitfsread.c [hunk 1]
```diff
@@ -238,19 +239,34 @@ cockpit_fsread_prepare (CockpitChannel *channel)
...
+      bytes = g_mapped_file_get_bytes (mapped);
+    }
+  else if (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NODEV))
+    {
+      gchar *contents;
+      gsize length;
+
+      g_clear_error (&error);
+      if (g_file_get_contents (self->path, &contents, &length, &error))
+        bytes = g_bytes_new (contents, length);
```
Comment (stefwalter) [inline src/bridge/cockpitfsread.c]: This should be g_bytes_new_take() so that the GBytes takes ownership of contents and length. This is both more efficient, and fixes a memory leak.

Comment (stefwalter) [inline src/bridge/cockpitfsread.c]: This is not very efficient, in certain corner cases, but pragmatically solves the actual issue. I think this is good enough until someone requires that those corner cases be fixed.

File: src/bridge/test-fs.c [hunk 2]
```diff
@@ -435,6 +435,31 @@ test_read_removed (TestCase *tc,
 }
 
 static void
+test_read_non_mmappable (TestCase *tc,
+                         gconstpointer unused)
+{
+  gchar *tag;
+  JsonObject *control;
+  const gchar *path = "/sys/power/state";
```
Comment (stefwalter) [inline src/bridge/test-fs.c]: The test should check that this file actually exists (it migth not in a container). If it doesn't exist, then the test should ```cockpit_test_skip()``` and ```return;```

File: src/bridge/cockpitfsread.c (hunk 1 above)
Comment (mvollmer) [inline src/bridge/cockpitfsread.c]: > This should be g_bytes_new_take()

Of course!

File: src/bridge/test-fs.c (hunk 2 above)
Comment (mvollmer) [inline src/bridge/test-fs.c]: Fixed.

**Comment by dperpeet** [comment]
> Ouch, I was confused. Watching files in /sys has nothing to do with mmap, but afaik they might require some special use of select. Or not. Just ignore me.

I don't know why we would need to watch that... I'm not even sure that changes there would be signalled.

**Comment by mvollmer** [comment]
> I don't know why we would need to watch that... I'm not even sure that changes there would be signalled.

Yeah, this is probably just some bad memory from the old days: I remember that files in /sys required a dedicated approach if you wanted to watch them with select, but inotify probably just works, and this is in fact what UDisks2 does, so we should be good.
